Implement A* pathfinding (partially)

roryokane opened the pull request:
The code compiles but does not pass all of the new tests like this. I still have more work to do.

The actual `AStarPathfinder` class is, I think, complete. It’s the glue code that I’m still working on.

I have made the design more complex by adding abstractions for the different possible ways to move on a grid: only horizontally and vertically, diagonally too, or diagonally but not allowing cutting through walls on corners. Those are the `GridMovementStyle`s, each of which contains an appropriate `DistanceHeuristic` and `NeighboringPositionFinder` for that style of movement. I needed a wrapper class because two movement styles share the same `DistanceHeuristic` but have different `NeighboringPositionFinder`s.

My next steps are implementing my TODO comments in [`NeighboringPositionFinder.scala`](https://github.com/ayai-ai/ayai-backend/blob/implement-pathfinder/src/main/scala/ayai/components/pathfinding/NeighboringPositionFinder.scala#L26-L38):

> ``` scala
> }).filterNot(position => ??? /* map(position) == NodeState.IMPASS */)
> ```
> - Should this take a map or a graph?
> - I do want to be able to look up the position just by a `Position`. That means it should define `apply()`.
> - But I can't use the `NodeMatrixGraphView` view for that, because that needs to call this to find neighbors.
> - So the object with an `apply()` definition will need to be an intermediate value between an `Array[Array[Node]]` and a `NodeMatrixGraphView`.
> - I think I will make a `NodeMatrix` class, and make `GraphFactory` return it.
> - There is one thing to check before I do: is it okay that we use `Position`s to index the grid? `GraphFactory` seems to avoid it, but I’m not sure whether that is because they thought it would be confusing to convert scaled positions to unscaled positions, or because a `Position` is only supposed to represent something specific to the game map, and should not be generic.
> - Look at the places `Position` is used in the code to see if it is ever used differently,
>   or if there are any comments saying not to.

roryokane committed:
Implement most of A* pathfinding

The code compiles but does not pass all of the new tests like this. I still have more work to do.

The actual AStarPathfinder class is, I think, complete. It’s the glue code that I’m still working on.

I have made the design more complex by adding abstractions for the different possible ways to move on a grid: only horizontally and vertically, diagonally too, or diagonally but not allowing cutting through walls on corners. Those are the GridMovementStyles, each of which contains an appropriate DistanceHeuristic and NeighboringPositionFinder for that style of movement. I needed a wrapper class because two movement styles share the same DistanceHeuristic but have different NeighboringPositionFinders.

My next steps are implementing my TODO comments in `NeighboringPositionFinder.scala`.

roryokane commented:
Christian, reload the page if you haven’t yet, to see the latest edit of the pull request description. It has more detail on what I still need to do.

chrisbenincasa commented on src/main/scala/ayai/components/pathfinding/NeighboringPositionFinder.scala:
```diff
@@ -0,0 +1,86 @@
+    // So the object with an apply() definition will need to be an intermediate value
+    // between an Array[Array[Node]] and a NodeMatrixGraphView
+    // I think I will make a NodeMatrix class, and make GraphFactory return it.
+    // There is one thing to check before I do: is it okay that we use `Position`s to index the grid?
```
If I'm understanding you correctly, I _don't_ think we can address a Graph (i.e. one returned by `GraphFactory.generateGraph`, an `Array[Array[Node]]`) by `Position`.This function, which is present in `GraphFactory.scala` seems to suggest this:

``` scala
private def convertPositionToGrid(position: Position, ratio: Float): Tuple2[Int, Int] =  {
    (math.round((position.x / ratio)), math.round((position.y / ratio)))
  }
```

That's used when setting a `(x, y)` coordinate when generating the graph to return. The giveaway there is the conversion from `Position` to the correct values using `ratio` which seems to be set from the `tileSize` of the map. If we _did_ want to use the a Graph like the one returned by `GraphFactory` (which I think you have here now), then we'll need to use a similar converter within `NodeMatrixGraphView` to convert from `Position` => `(Int, Int)` to address the matrix.

I think the graph is just a generalized version of the map that is not supposed to be dependent upon the tile size of the map. That said, I think it makes sense to use graph types for pathfinding. That way, pathfinding has no tie to tile size. In that case, we'd just pass around `Tuple2`s of positions which grid-based and convert them back to `Position`s at the end of the processing. What do you think?

chrisbenincasa commented on src/main/scala/ayai/components/pathfinding/NeighboringPositionFinder.scala:
```diff
@@ -0,0 +1,86 @@
+      Tuple2(-1, 0),
+      Tuple2(-1, -1),
+      Tuple2(0, -1),
+      Tuple2(1, -1)
```
You don't need to explicitly call `Tuple2` in Scala. You can just say `(1, 0)`

chrisbenincasa commented on src/main/scala/ayai/components/pathfinding/NeighboringPositionFinder.scala:
```diff
@@ -0,0 +1,86 @@
+ */
+class ChebyshevNeighboringPositionFinder extends NeighboringPositionFinder {
+  override def neighboringPositions(pos: Position): List[Position] = {
+    val coordinateDifferences = List(
```
Instead of defining this list each time the function is called, how about we keep it in a `private val` within the `class` as a constant?

roryokane commented:
I got `AStarPathfinder` to pass the following unit tests:
- `"Pathfinding" should "return None when there is no valid path"`
- `it should "be able to find a uniquely valid path"`
- `it should "find a maximally short path across almost-empty space"`
- `it should "return a global optimum even if it is locally suboptimal"`

The only test it is failing is this one:
- `it should "penalize slow terrain appropriately"`

The returned path is overly circuitous, and also goes through more `SLOW` nodes than it needs to.

On Tuesday, I will
1. Debug and fix that failing test
2. Search for and delete any stale TODO comments in my additions
3. Create two Git commits out of my changes – implementation of pathfinding, and some refactoring of `TileMap`
4. Push the commits to this pull request

---

I did avoid using `Position` in the pathfinding algorithms like you suggested. Instead of using a `Tuple2`, I created a `NodeMatrixPosition` class with more-descriptive `x` and `y` methods.

Do you know what a `Position` represents before its coordinates are divided by `tileMap.tileSize`? Why is such a division even necessary? Does it represent position within a tile, for fireballs that are smaller than the size of one tile or something?

I ask because I’m wondering if I can make the difference between `NodeMatrixPosition` and `Position` clearer. I changed `DistanceHeuristic.estimateDistance` to accept two `NodeMatrixPosition`s instead of two `Position`s, but it seems wasteful for that code to only work with positions specified in that format. It would make it harder for other parts of the program like Perception to use our distance algorithms.

roryokane committed:
Finish implementing AStarPathfinder

I use a different object from Position to represent locations on the map, to ensure the tile size ratio calculation is done when necessary. Instead of using a Tuple2 like the GraphFactory produced, I made a class NodeMatrixPosition whose coordinates can be accessed more easily with .x and .y.

The AStarPathfinder had some bugs that I fixed:
• I had forgotton to initialize the values of `cameFrom` and `costSoFar`.
• I have negated `priority` so nodes are visited in the correct order. In the sample implementation, lowest priority went first; in this code, highest priority goes first.
• I fixed the direction of the path to go from start to goal (the sample implementation is wrong), and made the path not include the start position.

I noticed and fixed a bug in GraphFactory. The positions were not bounds-checked correctly – it compared the x and y values to each other, instead of to the dimensions of the map. I also refactored GraphFactory some more.

roryokane commented:
Okay, the A* Pathfinder is done.

I will put the `TileMap` refactoring I mentioned in a separate pull request, since I noticed a bug in `TileMap` I want to fix first. This commit should work on its own.

#### What is left to do for pathfinding
-   Finishing `PathfindingSystem`, so that entities with a `Pathfinder` are able to call it from the entity-specific AI. For example, the system should make it possible to implement a character that walks around randomly, but when a player is within 10 tiles of the character, it walks towards the player using pathfinding.
  
  If we are low on time, we can hard-code the `PathfindingSystem` to do the moving of the character itself, but that would make pathfinding practically useless, and only good for demoing.
-   creating a demo character with a `Pathfinder` attached to it, and recording an in-game video of it moving along the path
-   (less important) adding unit tests for grid movement styles other than Manhattan, especially the `WallSensitiveChebyshevMovementStyle`, which has relatively complex logic
-   Implementing a `Pathfinder` for [D* Lite](http://en.wikipedia.org/wiki/D*#D.2A_Lite), as we said we would do in the Software Requirements Specification.
  
  I think this will be harder than A*, because the point of D* Lite is to save some state in between pathfindings so that the next one will be faster. The A* implementation doesn’t save any state, so `Pathfinder` and `PathfindingSystem` may have to be refactored to support that.

chrisbenincasa commented on src/main/scala/ayai/systems/ai/pathfinding/PathfindingSystem.scala:
```diff
@@ -25,18 +25,23 @@ class PathfindingSystem(actorSystem: ActorSystem) extends EntityProcessingSystem
-      pathfinding.findPath(map, position, position) match {
+      val movementStyle = new ManhattanMovementStyle
+      val graphView = new NodeMatrixGraphView(map, movementStyle)
+      val conversionRatio: Float = ???
```
This needs an actual number, what do you suggest?

roryokane replied:
The conversion _to_ `NodeMatrixPosition` happens in `GraphFactory`, and looks like this:

``` scala
val gridPosition = NodeMatrixPosition.fromPosition(p, tileMap.tileSize.toFloat)
```

So to convert back, we need to somehow look up which `tileMap` the current entity is on (which should be the same as the one previously passed to `GraphFactory`). Then just call `.tileSize.toFloat` on it.

chrisbenincasa commented on src/main/scala/ayai/components/pathfinding/Pathfinder.scala:
```diff
@@ -1,28 +1,85 @@
+    cameFrom(start) = start
+    costSoFar(start) = 0
+
+    while (! frontier.isEmpty) {
```
you can simplify to `while (frontier.nonEmpty)`, if you want.

chrisbenincasa commented on src/main/scala/ayai/systems/ai/pathfinding/PathfindingSystem.scala:
```diff
@@ -25,18 +25,23 @@ class PathfindingSystem(actorSystem: ActorSystem) extends EntityProcessingSystem
-//      println(s"Processing entity = ${e.uuid} for AStar at position $position")
+      //println(s"Processing entity = ${e.uuid} for AStarPathfinder at position $position")
 
       // TODO: replace with real target
```
How do we go about doing this?

chrisbenincasa commented:
So..from what I can tell, the server only sends JSON messages to the front-end in certain situations (initialization, room switch, etc). It looks like to get pathfinding to work, we'd have to setup a new message that gets sent over the WebSocket from the server to the client which indicates that the NPC should begin following a path.

chrisbenincasa commented:
Forget my last comment, can just pass the path down with the `update` json that gets sent over the wire.

roryokane committed:
Simplify `! … .isEmpty` to `.nonEmpty`

roryokane committed:
Define conversionRatio in PathfindingSystem

fixing the NotImplementedError

roryokane commented:
Pathfinding is not yet actually integrated into the game, but now this code at least doesn’t interfere with the rest of the game, so I am merging it in.

To truly implement pathfinding, [the steps in my earlier comment](https://github.com/ayai-ai/ayai-backend/pull/24#issuecomment-102245312) need to be completed.
